planner/core/preprocess.go

```diff
@@ -1377,6 +1381,11 @@ func (p *preprocessor) checkFuncCastExpr(node *ast.FuncCastExpr) {
 // handleAsOf tries to validate the timestamp.
 // If it is not nil, timestamp is used to get the history infoschema from the infocache.
 func (p *preprocessor) handleAsOf(node *ast.AsOfClause) {
+	readTS := p.ctx.GetSessionVars().TxnReadTS.PeakTxnReadTS()
+	if readTS > 0 && node != nil {
+		p.err = ErrAsOf.FastGenWithCause("can't use select as of while already set transaction as of")
+		return
```

Comment:
So an `AS OF` statement after `SET TRANSACTION READ ONLY AS OF` won't consume the `TxnReadTS` even if it causes an error? Is this expected?

Reply:
it is expected.